### PR TITLE
fix #855

### LINKED ISSUE #855: Ctrl-Z only working once
Hello,

This workflow make xonsh hang

```
$ vim
Ctrl-z
$fg
Ctrl-z
```

the first Ctrl-z is working (I go back to the shell). But after the second one vim is paused but what I type is ignored

This might be related to #853 since the final state is the same (I type in the terminal and only display what I type)

### PR DESCRIPTION
**Pull request by GuillaumeLeclerc**
This patch fix #855. But there are some other bugs I found with background tasks. 

I'll make some more issues about them. But at least this one is fixed, I really needed it to work properly

I hope my python style is improving. I never used it in a large scale project

### CODE REVIEW

**Comment by scopatz** [comment]
This looks reasonable to me.  @adqm, do you mind taking a look?  Also, can anyone think up a good way to test this kind of thing?  The more I think about it, the more I believe that we need to have these kinds of top-level integration tests.

**Comment by GuillaumeLeclerc** [comment]
Maybe we could start a xonsh process and write some commands to it. what I fear is that it will not be considered as a tty and will behave strangely

**Comment by scopatz** [comment]
Not saying that tests are blocking here, because it fixes the problem.

**Comment by scopatz** [comment]
Well, that would be one way... Python does have a PTY interface that we could use: https://docs.python.org/3/library/pty.html#module-pty

**Comment by GuillaumeLeclerc** [comment]
`bg` is not working anymore. Does anyone know if it's handled differently than `fg` ?

**Comment by GuillaumeLeclerc** [comment]
`jobs` is still buggy (i.e. never stops)

**Comment by adqm** [comment]
I can't take a look right now, but I should be able to later tonight.

It looks like job control got seriously messed up at some point in the past.  I'm going to assume/hope it was during one of my breaks, and not in the original code I put in :)

**Comment by GuillaumeLeclerc** [comment]
I fixed `jobs`. I also saw there is a difference in the behavior with other shells. `fg` always pop out the latest started task. In other shells it is usually the latest started OR resumed task.

If we want to implement this we will need one more field where we store the last time we resumed a task. Or use a FILO queue.

**Comment by GuillaumeLeclerc** [comment]
Status:

#### `Ctrl-z`

Works on commands, but not on python code

#### `fg` 

works on some processes:
- `vim`
- `node`
- `sleep`

but fails on others:
- `sh -c "sleep 5"`

when you fg with no task number then the behavior is not the same as other shells

#### `bg`

Not working at all. I haven't found out why yet

###### `jobs`

Seems fine now

**Comment by adqm** [comment]
Awesome, thanks!  I just ran this though a couple of paces.  Looks like the fix for #855 is working great!

`bg` and `fg` are still a little bit messed up when they aren't given args.  Sorry for the long wall of text, but here are some of the tests I ran:
```
hartz@thinky-515 ~ $ jobs
 [1] stopped: vim s.cpp & (25361)
hartz@thinky-515 ~ $ fg
*[1] running: vim s.cpp (25361)

 [1] stopped: vim s.cpp & (25361)
hartz@thinky-515 ~ $ gitk &
^Z
 [2] stopped: gitk & (25375)
hartz@thinky-515 ~ $ jobs
 [1] stopped: vim s.cpp & (25361)
 [2] stopped: gitk & (25375)
hartz@thinky-515 ~ $ fg 1
*[1] running: vim s.cpp (25361)
hartz@thinky-515 ~ $ jobs
 [2] stopped: gitk & (25375)
hartz@thinky-515 ~ $ bg
Cannot send nonexistent job to background.
hartz@thinky-515 ~ $ jobs
 [2] stopped: gitk & (25375)
hartz@thinky-515 ~ $ bg 2
*[2] running: gitk & (25375)
hartz@thinky-515 ~ $ jobs
*[2] running: gitk & (25375)
hartz@thinky-515 ~ $ fg
ValueError: max() arg is an empty sequence
```

I am happy to go in and make those changes, but if you want to since you're in the thick of it, that would be great, too!  I would vote for making all these related changes part of this same PR.

If you don't mind, we should also add this to the CHANGELOG as part of this PR (other docs shouldn't need updating since this is all just bugfixes).

**Comment by adqm** [comment]
> bg
> Not working at all. I haven't found out why yet

I had some success with `bg` just now.  Are there specific commands you've been trying that aren't working?

**Comment by GuillaumeLeclerc** [comment]
I'm going to sleep now, and I have a lot to do tomorrow so I'm not sure if I will have time. If you do then of course you can do it.

I think we may want to consider redoing this whole thing in a clean way with queues (It's a common practice in schedulers/job managers) and using the obj's to send the signal instead of bypassing it with raw OS calls. It might interfere at some point and causing #853. 

I think this is the only part left to make this shell robust :)

**Comment by GuillaumeLeclerc** [comment]
Would anyone have the courage to review a potential new job manager ? (I don't guarantee I will do it but if it won't be reviewed then I don't need to motivate myself)

**Comment by GuillaumeLeclerc** [comment]
Hey,

I did a massive refactor of the job manager.

It's working great from my point of view. Could you test it. `bg` should be working :) 

The PR is not done yet for two reasons:

- I'm not filling `__xonsh_active_job__` because I don't know what it should be, and what is it for
- I don't know windows so it is probably broken. Maybe someone could help me on this

**Comment by scopatz** [comment]
> Would anyone have the courage to review a potential new job manager ? (I don't guarantee I will do it but if it won't be reviewed then I don't need to motivate myself)

@GuillaumeLeclerc, in general, don't be worried about this; if you put in a PR, it will be reviewed, if not by any of the other great people with push rights, than I will get to it.  I prioritize looking at PRs more than other issues.

**Comment by scopatz** [comment]
@GuillaumeLeclerc - just skimming this very briefly, I think that this could really use some tests. We have been having a lot of problems with the job management stuff lately, and I think that part of the reason for that was that we have been fixing things as they go - but some of these fixes have ended up breaking other things.  

What we need is a good way to establish that the capabilities that we want are going to be maintained or improved by any changes that are made. I think this will also help the windows folks a lot (and maybe we should re-enable Appveyor?) because the test suite will tell us where to look if something breaks.

All that said, I *really* appreciate you taking this on.  The jobs management definitely needs a refactor to get everything working as expected. This refactor has been coupled with better testing in my mind, though.  Let us know if you have any questions about what this would entail.

Thanks again!

**Comment by GuillaumeLeclerc** [comment]
I'll try to write tests but I'll probably not have the time to do it today. 

I'll take a look at fish's repository and learn how they tested this part (and if they did so)

**Comment by adqm** [comment]
At a first glance, this looks really nice!  I'm happy to go through it with a fine-toothed comb when I have a chance, but that probably won't be until this coming weekend.

File: xonsh/jobs.py [hunk 1]
```diff
@@ -7,13 +7,18 @@
 import builtins
 from subprocess import TimeoutExpired
 from io import BytesIO
+from collections import deque
 
 from xonsh.tools import ON_WINDOWS
 
+RUNNING = 1
+STOPPED = 0
```
Comment (scopatz) [inline xonsh/jobs.py]: This is basically an enum.

File: xonsh/jobs.py [hunk 2]
```diff
@@ -97,43 +102,47 @@ def _give_terminal_to(pgid):
...
-                return
+        _continue(active_task)
 
         _, wcode = os.waitpid(obj.pid, os.WUNTRACED)
         if os.WIFSTOPPED(wcode):
-            job['bg'] = True
-            job['status'] = 'stopped'
             print()  # get a newline because ^Z will have been printed
-            print_one_job(act)
+            active_task['status'] = STOPPED
```
Comment (scopatz) [inline xonsh/jobs.py]: I think the string flags are more readable than integer flags.

Comment (scopatz) [inline xonsh/jobs.py]: Unless there is a reason for the integer ones I am not seeing...

**Comment by scopatz** [comment]
If @adqm is happy I am happy :).  Thanks for offering to put in some tests!

**Comment by adqm** [comment]
> I'm not filling \__xonsh_active_job__ because I don't know what it should be, and what is it for

It is very likely that `__xonsh_active_job__` can be completely replaced with `_get_next_task`.

> I don't know windows so it is probably broken. Maybe someone could help me on this

I can run some Windows tests this weekend.  Windows is kind of iffy with job control anyway (given that there isn't a clean way to suspend a process), but we can at least replicate some minimal piece of functionality.  Eventually, we'll want to dig deeper and figure out how to actually emulate job control there in full.

> I'm filling the start time for tasks but not end. Should it be done here ?

Personally, I think this is fine as is.

File: xonsh/jobs.py (hunk 2 above)
Comment (GuillaumeLeclerc) [inline xonsh/jobs.py]: It was only thinking about performance but I guess it's not relevant. I will change it back

Comment (scopatz) [inline xonsh/jobs.py]: Yeah, I think the performance hit is minimal enough for the gains in user readability.

**Comment by GuillaumeLeclerc** [comment]
__xonsh_active_jobs__ seems to be used outside `jobs.py` How is it used ?

**Comment by GuillaumeLeclerc** [comment]
fg and bg are not working with some processes, eg: eclipse. Maybe because we are sending the signal to a process instead of a process group.

What do you think ?

**Comment by scopatz** [comment]
Hi @GuillaumeLeclerc 

> xonsh_active_jobs seems to be used outside jobs.py How is it used ?

`__xonsh_active_jobs__` is used to set $PROMPT and $TITLE strings, which are pretty important, so it is probably not the best idea to remove all references to them.  See  https://github.com/scopatz/xonsh/blob/master/xonsh/environ.py#L1011

> fg and bg are not working with some processes, eg: eclipse. Maybe because we are sending the signal to a process instead of a process group.

Well they weren't working at all before, so this is probably an improvement :). I do think the active job thing needs to be cleaned up, though.

**Comment by adqm** [comment]
> \__xonsh_active_jobs__ is used to set $PROMPT and $TITLE strings, which are pretty important, so it is probably not the best idea to remove all references to them

That said, I think it should be possible to replace that with a call to `_get_next_task`, so long as we ensure that that task is actually running and not stopped.

I think this is definitely a big-time improvement already.  I'm happy to help out with this, but I'm pretty much booked until the weekend.

**Comment by scopatz** [comment]
> That said, I think it should be possible to replace that with a call to _get_next_task, so long as we ensure that that task is actually running and not stopped

I think that is better than `__xonsh_active_job__`. It is better to have an API call where we can, rather than using builtins.

**Comment by GuillaumeLeclerc** [comment]
The problem is the only time a job is active is when it's running. Therefore the shell does not have the terminal at this moment and we are not able to ask for the active job. As soon as the shell get the terminal back that means there are no more active tasks and it will always read None as active task.

Other thing (bad news). I look at the `fish-shell` tests and they is not a single test containing any of these built-ins:

- `fg`
- `bg`
- `jobs`

I'll try do investigate on other shells to check how they do this

**Comment by adqm** [comment]
Maybe we should just rename that variable then, but still keep it around?  We use it to set the title of the terminal, which I find nice, but we don't have to actually use it for anything more meaningful than that.  I'm happy to help work out the details if need be.

Testing these things (any interactive command, really) is going to be hard.  I know it would be desirable, but I'm not sure it's feasible.

**Comment by GuillaumeLeclerc** [comment]
I'll try to fix the title then.

Tcsh does not test this either. The only one I found who has real tests about job control is bash. It seems they are just running scripts and comparing the output with a reference.

**Comment by GuillaumeLeclerc** [comment]
I found out the problem with job control and complex tasks !!! :D 

We were sending signals to the the main process only, not the process and all its children. that's why we were not able to return to complex processes (firefox, ecplise, `sh some_script.sh`). This was because their children were still in the interrupted state.

So now we are sending the required signals to the whole process group and it working perfectly fine. I tested bash scripts and the Scala IDE.

**Comment by adqm** [comment]
Ah, yes; that makes good sense!  Is this ready for a careful look, then?  If so, I'll go through and do a thorough review tonight.

Very glad to have this; thanks!

**Comment by GuillaumeLeclerc** [comment]
Not yet, I still have to fix the "start in background" feature. It's not working for me at the moment

**Comment by GuillaumeLeclerc** [comment]
I dug into this problem this evening and it is a problem with the parser. `built_in.py` receive the commands without their trailing `&`. I tried looking at the parser but I could not understand anything. I was expecting a formal grammar or something like this but I got completely lost.

I think the fix of the parser should not be part of this PR in my opinion.

The only thing left is reviewing the whole PR and someone to take care of the windows part. I don't have a single machine running it so It will be complicated for me. If nobody can do it then I'll try to get the motivation and install a virtual machine.

**Comment by adqm** [comment]
I should be able to look into those things tonight or tomorrow.

@scopatz, is there a hard deadline for the release?

**Comment by scopatz** [comment]
Sorry I missed this @adqm. I anticipate that we might need to do a v0.3.1 release next week, just to clean things up a bit.

**Comment by adqm** [comment]
Ah, no worries.  I'll still try to test this thoroughly tonight.

**Comment by scopatz** [comment]
Thanks!  I think we are almost certainly going to want a v0.3.1 release now.  That 'best' thing is pretty annoying.

File: xonsh/jobs.py [hunk 3]
```diff
@@ -7,13 +7,18 @@
...
+RUNNING = 1
+STOPPED = 0
+
+tasks = deque()
 
 if ON_WINDOWS:
-    def _continue(obj):
-        return None
+    def _continue(job):
+        job['status'] = 'running'
```
Comment (adqm) [inline xonsh/jobs.py]: should this be `RUNNING`, for the sake of consistency with the other pieces?

**Comment by adqm** [comment]
I think this all looks great, @GuillaumeLeclerc!  Thanks again for putting it in.

I just put in #962, which builds on this and (hopefully) resolves some of the remaining issues.  Hoping to get this and those new changes merged in at the same time.

**Comment by scopatz** [comment]
Hi @GuillaumeLeclerc - I believe that these changes have been included by @adqm  in #962 (though he rebased so this didn't autoclose.) Please feel free to close this.

**Comment by scopatz** [comment]
Thanks again for putting it in.

**Pull request closed without merging**